### PR TITLE
UCDC/Borovenskiy/feature/api to register auth users to edx

### PR DESCRIPTION
**Pull request by NikolayBorovenskiy**
**Description:** PR creates edx plugin to cooperate with ucdc portal via restful APIs.
Consist of:
 - plugin skeleton,
 - API user registration,
 - swagger doc for the API,
 - tests for the API.

**Youtrack:** https://youtrack.raccoongang.com/issue/UCDC-45

### CODE REVIEW

File: common/lib/ucdc_edx_api_plugin/docs/api/user.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+          properties:
+            username:
+              type: string
+              description: The user name.
+            email:
+              type: string
+              description: The user email.
+            uuid:
+              type: string
+              description: User ID for UCDC potral. Also is as OAuth privider ID since UCDC potral is OAuth privider for Edx.
```
Comment (Masergik) [inline common/lib/ucdc_edx_api_plugin/docs/api/user.yml]: Mistyped words: `potral` and `privider`

Comment (NikolayBorovenskiy) [reply common/lib/ucdc_edx_api_plugin/docs/api/user.yml]: Thx! My fault.

File: common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/serializers.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    auth_uid = serializers.SerializerMethodField(read_only=True)
+
+    class Meta:
+        model = User
+        fields = ["id", "username", "email", "auth_uid"]
+
+    @staticmethod
+    def get_auth_uid(obj):
+        qs = UserSocialAuth.objects.filter(user_id=obj.id)
+        if qs.count() == 1:
```
Comment (idegtiarov) [inline common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/serializers.py]: Please pay attention to this part of the code.

File: common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/views.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    authentication_classes = (OAuth2AuthenticationAllowInactiveUser,)
+    permission_classes = (ApiKeyHeaderPermission, AnonPermissionOnly)
+
+    def post(self, request, *args, **kwargs):
+        params = request.data.copy()
+        user_serializer = UserRegisterSerializer(data=params)
+        if user_serializer.is_valid(raise_exception=True):
+            user = user_serializer.save()
+        UserProfile.objects.create(user=user, allow_certificate=True)
+        params["provider"] = OAuth2ProviderConfig.objects.first().backend_name
```
Comment (idegtiarov) [inline common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/views.py]: Let's add a comment that `provider` should be only one or first in the admin dashboard.

File: requirements/edx/base.txt [hunk 4]
```diff
@@ -260,3 +260,6 @@ xblock==1.2.1
 zendesk==1.1.1
 django-tinymce==2.8.0
 django-hvad==1.8.0
+
+# UCDC integration
+-e common/lib/ucdc_edx_api_plugin
```
Comment (idegtiarov) [inline requirements/edx/base.txt]: requirements/edx/local.txt seems to be the better place to install the package.

File: common/lib/ucdc_edx_api_plugin/requirements/dev.txt [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Requirements, needed for the development.
+
+-r base.txt
+
+#Tools for testing and style checking
```
Comment (idegtiarov) [inline common/lib/ucdc_edx_api_plugin/requirements/dev.txt]: Please add space after `#` ))))

Comment (NikolayBorovenskiy) [reply common/lib/ucdc_edx_api_plugin/requirements/dev.txt]: Add a one ))

**Comment by NikolayBorovenskiy** [comment]
Seems to me that most of all comments are covered. Look forward to a final review.

File: common/lib/ucdc_edx_api_plugin/docs/api/user.yml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+          description: The user ID.
+        username:
+          type: string
+          description: The user name.
+        email:
+          type: string
+          description: The user email.
+        auth_uid:
+          type: string
+          description: The OAuth provider user ID.
```
[the remark this answers is not in the document]
Comment (NikolayBorovenskiy) [reply common/lib/ucdc_edx_api_plugin/docs/api/user.yml]: done

File: common/lib/ucdc_edx_api_plugin/requirements/dev.txt [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+coverage
```
[the remark this answers is not in the document]
Comment (NikolayBorovenskiy) [reply common/lib/ucdc_edx_api_plugin/requirements/dev.txt]: done

File: common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/permissions.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
+from rest_framework import permissions
+
+
+class AnonPermissionOnly(permissions.BasePermission):
+    message = 'You are already authenticated. Please log out to try again.'
+    """
```
[the remark this answers is not in the document]
Comment (NikolayBorovenskiy) [reply common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/permissions.py]: done

File: common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/serializers.py (hunk 2 above)
Comment (NikolayBorovenskiy) [reply common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/serializers.py]: done

File: common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/views.py (hunk 3 above)
Comment (NikolayBorovenskiy) [reply common/lib/ucdc_edx_api_plugin/ucdc_edx_api/api/views.py]: done

File: requirements/edx/base.txt (hunk 4 above)
Comment (NikolayBorovenskiy) [reply requirements/edx/base.txt]: done

File: requirements/edx/custom.txt [hunk 9]
```diff
@@ -1,3 +1,6 @@
 # RACCOONGANG
 sentry-sdk==0.7.8
 # RACCOONGANG
+
+# UCDC integration
+-e lms/lib/ucdc_edx_api_plugin
```
Comment (idegtiarov) [inline requirements/edx/custom.txt]: I still no agree, local.in looks more attractive to put local libs with the own setup files.

Comment (NikolayBorovenskiy) [reply requirements/edx/custom.txt]: Perhaps local is a more understandable and convenient place. But I had a conversation on this subject with DevOps engineer Oleksii Gnedin. He made it clear to me from which files the environment on Staging or Production is deployed. Always from .txt and not from local.txt (in) for sure. I have no expertise here and I trust my colleagues. Alex was very convincing 🙂

File: requirements/edx/local.in [hunk 10]
```diff
@@ -10,3 +10,6 @@
 -e common/lib/xmodule
 
 -e openedx/core/lib/xblock_builtin/xblock_discussion
+
+# UCDC integration. Plugin that describes the REST-full api through which UCDC portal interacts with the EDX platform.
+-e common/lib/ucdc_edx_api_plugin
```
Comment (idegtiarov) [inline requirements/edx/local.in]: Please add a trailing empty line, other LGTM

**Pull request merged**